### ISSUE TITLE
Multiple optimizers using optax

### ISSUE THREAD

**Issue by DavidSlayback** [deepmind/optax#59]
Sorry if this already exists in optax as a feature, but how would you go about making a multi-optimizer (similar to [Flax optim](https://flax.readthedocs.io/en/latest/_modules/flax/optim/base.html#MultiOptimizer)) that could use different learning rates for different parts of a network? 

Specifically, I'm have a full model in Haiku with one learning rate for most of the parameters, but different learning rates for specific subsets. I can partition the params appropriately and create a separate optimizer for each subset, but ideally I'd like to maintain the simplicity of a single set of params and optimizer_state. Is there a common approach to this?

**Comment by n2cholas** [comment]
One possible way to do this is by combining `chain` and `masked`. Here's an example using Haiku where you apply different optimizers to the weights vs biases:

```python
import jax
import jax.numpy as jnp

import optax
import haiku as hk

def forward_fn(x):
  return hk.Sequential([
    hk.Linear(10),
    hk.Linear(10),
  ])(x)
forward = hk.transform(forward_fn)

params = forward.init(jax.random.PRNGKey(0), jnp.ones([1, 1]))

# a pytree with the same structure as params with boolean values at the leaves
is_bias_mask = hk.data_structures.map(lambda mname, name, val: name != 'b', params)
not_bias_mask = jax.tree_map(lambda x: not x, is_bias_mask)

optimizer = optax.chain(
    # only applies to leaves where is_bias_mask is True:
    optax.masked(optax.sgd(learning_rate=0.1), is_bias_mask), 
    # only applies to leaves where not_bias_mask is True:
    optax.masked(optax.adam(learning_rate=0.01), not_bias_mask)) 

state = optimizer.init(params)
...
```

Once #60 is merged, it will be possible to use `masked` without knowing the structure of the params pytree in advance:

```python
from functools import partial

# mask_fn returns a mask pytree given the parameters pytree as input
is_bias_mask_fn = partial(hk.data_structures.map, lambda mname, name, val: name != 'b')
not_bias_mask_fn = partial(hk.data_structures.map, lambda mname, name, val: name == 'b')

optimizer = optax.chain(
    optax.masked(optax.sgd(learning_rate=0.1), is_bias_mask_fn),
    optax.masked(optax.adam(learning_rate=0.01), not_bias_mask_fn))
```

Does that work for your use case?

**Comment by DavidSlayback** [comment]
Yeah, that works well for my case, I can create masks for each param group, then do a chain entry for each. I appreciate the help, thank you! Any idea when #60 might be merged? It already looks like I might be able to just copy it to my local repo for now

**Comment by n2cholas** [comment]
Great! #60 was merged a few hours ago, so you should be able to use it now 😊.

**Comment by n2cholas** [comment]
As of #132, this is much simpler: https://optax.readthedocs.io/en/latest/api.html?highlight=multi#optax.multi_transform
